### ISSUE TITLE
Make opting out of I/O managers not rely on `-> None` return annotation 

### ISSUE THREAD

**Issue by jamiedemaria** [dagster-io/dagster#15590]
### What's the use case?

Right now, in order to fully opt out of using I/O managers, you need to have the `-> None` return type annotation:

```python
@asset 
def no_io() -> None:
     make_table_in_db()
```

if you do this instead:
```python
@asset 
def no_io():
     make_table_in_db()
```

You will see different behavior depending on the I/O manager (pickling a None object for file system I/O managers, raising an error for DB I/O managers, etc). Additionally, there will be messages logged about the I/O managers being used

Depending on which I/O manager you use, this may not be a huge problem (maybe you dont care that the pickled Nones are getting saved to files), but I think we should explore bailing out of the I/O manager path if the return type is None regardless of the return type annotation. 

However, bailing out of the I/O manager on a returned `None` could also have some consequences (and maybe breaking changes) for users who are intentionally returning `None` from their assets and want that value loaded in a downstream asset. 

Contrived example:
```python
@asset 
def returns_none():
    return None

@asset 
def expects_none(returns_none):
    assert returns_none is None
```

We'd also need to explore how this affects conditional materialization

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by sryza** [comment]
I think this would be great. Thoughts on how we can do it without breaking compatibility?

**Comment by jamiedemaria** [comment]
one idea is to have the I/O managers supply a `None` if the upstream asset doesn't have a stored value. But that could cause other issues if a user materialized a random downstream asset without materializing the upstreams first. i think it's worth prototyping

**Comment by sryza** [comment]
> But that could cause other issues if a user materialized a random downstream asset without materializing the upstreams first.

Yeah. A potential route to limiting the negative impact would be checking to see if an upstream materialization event exists.
